### ISSUE TITLE
Dependency information for FMUs for ModelExchange and CoSimulation

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/fmi-standard#380]
**Reported by cbertsch on 21 Jun 2016 12:43 UTC**
(Ticket by Andreas Nicolai <andreas.nicolai@tu-dresden.de> which could not be entered due to technical reasons of TRAC)

It is currently unclear, how dependency information (model structure) has to be specified in modelDescription.xml file when an FMU is exported for both ModelExchange and CoSimulation.

Consider the following example:

A model takes an input `u_1` and computes and output `y_1`. The model consists of a single ordinary differential equation for variable `x[0]` and its derivative `der(x[0])`.

For example:
```
  u_1 : Heat load
  y_1 : Temperature

  x   : energy density
  der(x[0]) : time derivative of energy density

  equation:
    der(x[0]) = u_1;
    x[0] = y_1*heat_capacity;
```

If this model were exported as ModelExchange FMU, the dependency graph
(1) would look like:

```
   u_1 --> der(x[0])
   x[0] --> y_1
```

However, if the same model were exported as CoSimulation FMU, the internal variable is no longer visible/used by the CoSimulation master. There will be an indirect dependency between y_1 and u_1, but only through calls of doStep() which could be represented by the graph (2):

```
   u_1 --> y_1
```

Since the current FMI standard only allows specifying a single model structure, both variants (1) and (2) cannot be placed in the same modelDescription.xml file.

From the perspective of a CoSimulation master, the graph (2) could be deducted from graph (1).
However, graph (1) cannot be reproduced from graph (2), which means that for ModelExchange FMUs the first graph representation is mandatory.

It should be clarified in the standard how such dependency information shall be given when an FMU is exported for both ME and CS.

Also, for CoSimulation it is not clear how direct algebraic relations, for example from an equation `u_2 == y_2`, can be distinguished from indirect dependencies (that are indirectly given by the model equations and only appear through doStep() call).
For master algorithms that need to distinguish between these dependency types, the standard should also clarify how direct dependencies are marked.

----

Migrated-From: https://trac.fmi-standard.org/ticket/380

**Comment by modelica-trac-importer** [comment]
**Comment by aviel on 10 Jan 2017 13:41 UTC**
According to the 2.0 specification, p. 55:
```
The structure of the model is defined in element “ModelStructure”
within “fmiModelDescription”.
This structure is with respect to the underlying model equations,
independently how these model equations are solved.
```
So, only the variant (1) seems relevant for both model exchange and co-simulation.

**Comment by modelica-trac-importer** [comment]
**Modified by andreas&#46;junghanns on 6 Feb 2017 10:53 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by karl&#x2e;wernersson on 15 Jun 2017 12:14 UTC**
I agree with Antoine.

In addition it is possible to to expose these indirect dependencies by exposing the derivative in the modelStructure for co-simulation. 

/Kalle

**Comment by KarlWernersson** [comment]
proposal: close with wont fix

**Comment by KarlWernersson** [comment]
add extra simple
exaple
y=x
der(x)=u

dependencies y on x, and x on u, no dependencies y on u.

**Comment by chrbertsch** [comment]
Closing as included in merged PR #579
